contracts/NFTVesting/ConvexVestingNFTTimeLock.sol

```diff
@@ -0,0 +1,201 @@
+// SPDX-License-Identifier: MIT
+// OpenZeppelin Contracts (last updated v4.5.0) (token/ERC20/utils/TokenTimelock.sol)
+// SpartanLabs Contracts (NFTVesting)
+
+pragma solidity ^0.8.0;
+import "./IERC721.sol";
+
+/**
+ * @dev A single NFT holder contract that will allow a beneficiary to extract the
+ * NFT after a given vesting start time.
+ *
+ * After the vesting start time, the discount will start to accumulate for the locker linearly.
+ *
+ * Developers would have to perform the following actions for the locking of NFT:
+ * Deploy -> Approve -> Transfer
+ *
+ * Note that in order for discount in ETH to be valid, ETH must first be sent to this contract upon token locking.
+ */
+contract ConvexVestingNFTTimeLock {
+    // ERC721 basic token smart contract
+    IERC721 private immutable _nft;
+
+    // ERC721 basic token ID of contract being held
+    uint256 private immutable _tokenId;
+
+    // beneficiary of token after they are released
+    address private immutable _beneficiary;
+
+    // timestamp when token release is enabled and when discount starts to vest. Also known as cliff period.
+    uint256 private immutable _vestingStartTime;
+
+    // Max discount allowed for a token in percentage
+    uint256 private immutable _maxDiscountPercentage;
+
+    // Growth rate for vesting. M in MX^exponent
+    uint256 private immutable _growthRate;
+
+    // Exponent for vesting. exponent in MX^exponent
+    uint8 private immutable _exponent;
+
+    // Events
+    event EthReceived(address indexed sender, uint256 amount);
+
+    /**
+     * @dev Deploys a timelock instance that is able to hold the token specified, and will only release it to
+     * `beneficiary_` when {release} is invoked after `vestingStartTime_`. The vesting start time is specified as a Unix timestamp
+     * (in seconds).
+     *
+     *  The discount accumulation for beneficiary is based off a convex model y = mx^exponent
+     *  The developer would have to send ETH to this contract on contract deployement for discount to be applied.
+     *
+     */
+    constructor(
+        IERC721 nft_,
+        uint256 tokenId_,
+        address beneficiary_,
+        uint256 vestingStartTime_,
+        uint256 maxDiscountPercentage_,
+        uint256 growthRate_,
+        uint8 exponent_
+    ) {
+        require(
+            vestingStartTime_ > block.timestamp,
+            "Timelock: vesting start time is before current time"
+        );
+        require(
+            maxDiscountPercentage_ <= 100,
+            "TimeLock: max discount is greater than 100%. Please use a valid maxDiscountPercentage."
+        );
+        require(
+            address(this).balance > 0,
+            "Time:Lock: Eth should be sent to contract before initialization"
+        );
```

Comment:
This part seems odd - we're both allowing users of this contract to deposit an arbitrary amount of ETH and setting an arbitrary maxDiscountPercentage

2 arbitraries creates a possibility of mistakes

What do you think about allowing users to only define one?

For e.g. if they send in 1 ETH, that's the max discount and max discount is always 100% over time, otherwise there will always be remnants in this contact, especially if they send in 1 ETH and the max discount percentage is only 50%?

Reply:
hmm i think that is a better implementation. setting max discount to 100% so that eventually all eth sent to the contract would be claimed.

Reply:
Right and that would also make it such that the part that is paid and not discounted still goes to the developers - otherwise the developers would then only get their raised funds when they withdraw from this contract at the end of the vesting - which would be odd